Introduce links to CDEvents (#139)
cdevents/spec: links.md

Author: The technical implementation should be left out, but I think mentioning and describing the purpose of the links service can remain. I'll go ahead and move that, and link to it

Reviewer: @xibz , I believe that the spec repo should be used for the CDEvents protocol specification only. Since the PR currently proposes a separate dedicated links protocol, I believe it should be kept outside the CDEvents spec repo. The updates to the event schemas (e.g. chain_id) should of course be here, together with documentation updates in https://github.com/cdevents/spec/blob/main/spec.md#cdevent-context for the chain_id.

Reviewer 2: > The updates to the event schemas (e.g. chain_id) should of course be here, together with documentation updates in https://github.com/cdevents/spec/blob/main/spec.md#cdevent-context for the chain_id.

Yeah, I agree, the `chain_id` should be added to the context.

> Since the PR currently proposes a separate dedicated links protocol, I believe it should be kept outside the CDEvents spec repo. 

We discussed at some point having them inlined in CDEvents as an alternative to sending them on a separate channel - it's not the current proposal, but perhaps we could keep them for now in a separate repo? 
We could have an extensions folder or something along those lines?

# Connecting Events - Links Proposal

## Abstract

This proposal will outline how to connect individual CDEvents to eachother.
Right now there's no way of associating events to one another without needing

> to backtrack across certain context attributes, eg [id](https://github.com/CDEvents/spec/blob/main/spec.md#id-context).

Reviewer 2: The event `id` is unique to each event, I don't think it can be used to backtrack across events.
The [subject `id`](https://github.com/CDEvents/spec/blob/main/spec.md#id-subject) is a better candidate for that since the same subject can be part of multiple events, however, connecting events only through that can be challenging and in some cases impossible.

Reviewer: @afrittoli , are you satisfied with the current wording in this PR regarding this comment now?

Reviewer 2: @e-backmark-ericsson @xibz this is not critical, but as I mentioned in my previous comment, we cannot backtrack today through the context id, since that's unique to a specific event. 
The only way to backtrack today is through the subject id, so I would update the link to https://github.com/CDEvents/spec/blob/main/spec.md#id-subject

> There is limitations, however, in that we don't know when an event begins

Reviewer 2: Some subjects have `start` and `end` predicates (like `TaskRun` or `TestCaseRun`) - the event themselves do not begin or end, but the activity they describe may do; those activities may be part of a larger "workflow" that may begin and end too.

Reviewer: @afrittoli , are you satisfied with the current wording in this PR regarding this comment now?

nor finishes.

// sentence needs to be better
This proposal will outline a new approach that will allow for
the semantics of starting of a span and finishing.

## Semantics

This section will line out various definitions to ensure there are no
assumptions being made when we talk about linking events

* **CI** - Continuous integration
* **CD** - Continuous delivery
* **CDEvents Span** - A CDEvents span is an end to end representation of the

>   CI/CD lifecycle of some artifact 

Reviewer: Well 'end to end' here seems to have a quite limited scope. As I understand the proposal it can only cater for a 'fan-out' scenario, where the span is expected to be declared when a source change is being created for example, and then used in all succeeding activities down the line until a delivery is deployed on a customer site. What about issues/bugs/features that might precede such a source change? What span should that belong to?

I'm not sure the term `artifact` is correct here either. It's not the same as used when sending artifact.packaged/published events, or?

Author: action item - Specifically address confusion around how users can define their own custom `CDEvents Span` lifecycles.

* **Link** - A link is some relation to some thing, eg event, grouping, etc.
* **Global ID** - A global ID is some overarching ID for a given event.

## Goals

The biggest challenge in this whole process is ensuring that links can be
retrieved quickly and efficiently, while also providing the necessary metadata
and structure to give a detailed picture of how things occurred.

1) Provide a way of quickly retrieving all related links
2) Keep link data structured small and simple
3) Scalable

## Design

This section will propose two designs. The first being how individual events
can be linked to one another and be described in a way where it represents the
complete picture of the whole CI/CD span. The second portion will address the
goal of scalability.

### New Field(s)

When determining a course of action, it is generally important to consider what
the parent event had done. To cater to this very common use case, we will
introduce a new field called `parent` in each event.

The new `parent` field will be an array of the immediate preview parent(s)
events.

> `global_id` will also be added to the context of events which will allow for

Reviewer: I see issues with trying to define a global_id that should be used in all possible upstream/downstream events in a flow. See my comment on the CDEvents Span above for example.
Could an alternative be to introduce a context.spans array instead, where each item could be a {name=my_name, value=my_value} dict? That way an event could belong to multiple spans. Filtering on a dict within an array is not easily indexed in a generic DB service, so the speed gain might be less in that case, unless the DB could do some smarter index setting than just indexing the raw event data.

Reviewer 3: Will all events linked to each other have the same `global_id` ? 

If so, I wonder if a better name for this can be something like
- family_id or 
- span_id or
- link_id

Author: > Will all events linked to each other have the same global_id ?

Generally they will, but they could have different `global_id`.

Reviewer: I believe this comment is outdated now, based on changes done to the PR lately.

querying all links from a given global id.

The new parent field will be an added optional field, but the `global_id` will be
a required field in **all** cdEvent types looking something like:

```json
{
  "context": {
    "version": "0.3.0-draft",
    "global_id": "00000000-0000-0000-0000-000000000001", // new field
    "id": "505b31c2-8bc8-47b3-a1a0-269d7a8530ac",
    "source": "dev/jenkins",
    "type": "dev.cdevents.testsuite.finished.0.1.1",
    "timestamp": "2023-03-20T14:27:05.315384Z"
  },
  "subject": {
    "id": "MyTestSuite",
    "source": "/tests/com/org/package",
    "type": "testSuite",
    "content": {}
  },

>   "parent": [ // new proposed field here

Reviewer 2: If we include the whole parent event(s), a producer must always prune previous parents (if any) or else the structure will grow indefinitely.
What is the benefit of including the whole event body as opposed to the `context.id` alone?

Author: action item - add image to better illustrate parent

Source Change -> Triggers Tests -> (Pass) -> Uploads Artifact
CD would consume (Upload Artifact) CDEvent which would have the `Source Change` as parent.

## Issues

* Source changes can be large. Meaning we may need a subset of the parent rather than the whole thing.
* Clarify that it is an immediate logical grouped root parent

Reviewer: I believe this comment is outdated now, given the changes done to this PR lately.

    {
      "context": {
        "global_id": "00000000-0000-0000-0000-000000000001",
        "version": "0.4.0-draft",
        "id": "271069a8-fc18-44f1-b38f-9d70a1695819",
        "source": "/event/source/123",

>         "type": "dev.cdevents.change.merged.0.1.2",

Reviewer: @dan-han-101 , is the current state of this PR understandable regarding the above comment now?

>         "type": "dev.cdevents.change.merged.0.1.2",

Reviewer: The PR has evolved quite a bit since this comment was made, so I don't think it is valid anymore. @dan-han-101 , please state if you don't agree within a week or so, otherwise we'll consider this comment resolved.

      ...
    ]
  }
  ```

* `/links`
  * `GET`
  * Returns a list of links

  ```
  REQUEST
  ---
  {
    "pagination_token": "<some token>"
  }
  ```

* `/links/event`
  * `POST`
  * Adds a CDEvent to the links backend

  ```
  REQUEST
  ---
  {
    "event": {
      "context": {
        "version": "0.3.0-draft",
        "id": "505b31c2-8bc8-47b3-a1a0-269d7a8530ac",
        "source": "dev/jenkins",
        "type": "dev.cdevents.testsuite.finished.0.1.1",
        "timestamp": "2023-03-20T14:27:05.315384Z"
      },
      "subject": {
        "id": "MyTestSuite",
        "source": "/tests/com/org/package",
        "type": "testSuite",
        "content": {}
      }
    },
    "links": [ // optional
    ]
  }
  ```

### Scalability

Scalability is one of the bigger goals in this proposal and we wanted to ensure
fast lookups. This section is going to describe how the proposed links format
will be scalable and also provide tactics on how DB read/writes can be done.

The purpose of the `global_id` was to ensure very fast lookups no matter the
database. We could say that only graph DBs could be used to do a full span
lookup without a `global_id` but that poses two problems:

> * Slower lookups as the graph DB needs to backtrack to find the full span

Reviewer: Yes, slower than having a `global_id` in all related events, but it hasn't been a major show stopper for our use of simple uuid links in the Eiffel protocol so far. We have considered caching upstream queries towards our DB but not found the business case for that urgent enough to be implemented yet.

Author: action item - `global_id` could potentially be created on the links service itself. Look into this.

> * Limits to only graph DBs

Reviewer: Not true. In Eiffel we've used MongoDB since day one and still haven't seen a business case to replace it with a graph DB, even though we constantly perform "graph" traversals on the data, both upstream and downstream, based on the `links` and globally unique `id` uuid's used by the Eiffel event protocol.

Author: action item -  absolutely not true, but needs to reworded or simply removed

finishes.

## Semantics

This section will line out various definitions to ensure there are no
assumptions being made when we talk about linking events

* **CI** - [Continuous integration](https://bestpractices.cd.foundation/learn/ci/)
* **CD** - [Continuous delivery](https://bestpractices.cd.foundation/learn/cd/)
* **Span** - A span is an end to end representation of all activities performed
  in the CI/CD lifecycle of some entity
* **Link** - A link is an object containing relations between two CDEvents.
* **Global ID** - A global ID is some overarching ID for a given span.

## Goals

The biggest challenge in this whole process is ensuring that connected events
can be retrieved quickly and efficiently, while also providing the necessary
metadata and structure to give a detailed picture of how things occurred.

1) Provide a way of quickly retrieving all related links
2) Keep link data structured small and simple
3) Scalable

## Use Cases

This section will go over a few use cases and explain how this approach can be
used to solve for each particular case.

### 1. Fan Out Fan In

The fan out fan in use case is an example where a system may make parallel

> requests (fan out), and merge back to some other or the very same system (fan in)

Reviewer 4: Loving this proposal! Do you want us to review spelling? I am a terrible speller, ask anyone I work with 🤣, but if we see any small changes do you want us to make suggestions? 
```suggestion
requests (fan out), and merge back to some other or the very same system (fan in).
```

Author: Sure make suggestions! I went ahead and ran it through a spell checker (installed a vim plugin to do that). So it caught a few mistakes

Reviewer: I believe this should be considered resolved by now

Let us assume we have 3 system in our CI/CD environment. A continuous
integration environment, which we will call CI system, that runs tests and
builds artifacts, an artifact store that receives artifacts from the CI system,
and lastly the CD system which consume these artifacts shown in the diagram
below.

> ```

Reviewer 4: Haha love this over the original image! What did you say was the program `vim box`?

	"parent_ids": ["11111111-0000-0000-0000-000000000000"],
    "source": "/event/source/123",
    "type": "dev.cdevents.pipelinerun.queued.0.1.1",
    "timestamp": "2023-03-20T14:27:05.315384Z"
  },
  "subject": {
    "id": "mySubject123",
    "source": "/event/source/123",
    "type": "pipelineRun",
    "content": {
      "pipelineName": "myPipeline",
      "url": "https://www.example.com/mySubject123"
    }
  }
}
```

**Pipeline Started CDEvent**
```json
{
  "context": {
    "version": "0.4.0-draft",
    "id": "33333333-0000-0000-0000-000000000000",
	"global_id": "00000000-0000-0000-0000-000000000000",
	"parent_ids": ["22222222-0000-0000-0000-000000000000"],
    "source": "/event/source/123",
    "type": "dev.cdevents.pipelinerun.started.0.1.1",
    "timestamp": "2023-03-20T14:27:05.315384Z"
  },
  "subject": {
    "id": "mySubject123",
    "source": "/event/source/123",
    "type": "pipelineRun",
    "content": {
      "pipelineName": "myPipeline",
      "url": "https://www.example.com/mySubject123"
    }
  }
}
```

With these three separate CDEvents we can see that these all belong to a single
span of `00000000-0000-0000-0000-000000000000`, and that the relation is as follows

> ```

Reviewer 4: The diagram makes sense after reviewing it a few times, how the parent is inferred from downstream events. Reading right-to-left I had to review it a few times, would a top-to-bottom design make more sense? It might not but just my thoughts.

Reviewer: @elmsdata , are you satisfied based on the updates made to this PR lately?

This section will line out various definitions to ensure there are no
assumptions being made when we talk about linking events

* **CI** - Continuous integration
* **CD** - Continuous delivery
* **CDEvents Span** - A CDEvents' span is an end to end representation of the
  CI/CD lifecycle of some entity
* **Link** - A link is some relation to some thing, eg event, grouping, etc.
* **Global ID** - A global ID is some overarching ID for a given span.

## Goals

The biggest challenge in this whole process is ensuring that links can be
retrieved quickly and efficiently, while also providing the necessary metadata
and structure to give a detailed picture of how things occurred.

1) Provide a way of quickly retrieving all related links
2) Keep link data structured small and simple
3) Scalable

## Design

This section will propose two designs. The first being how individual events
can be linked to one another and be described in a way where it represents the
complete picture of the whole CI/CD span. The second portion will address the
goal of scalability.

### New Headers

To allow for connecting events and proper propogation, we will add two new HTTP
headers:
* `X-CDEVENTS-GLOBAL-ID`

> * `X-CDEVENTS-PARENT-IDS`

Reviewer 2: CDEvents in itself does not rely on HTTP in any way. 
We could define fields that will render into HTTP headers in certain bindings (i.e. CloudEvents in binary mode).
Note that CloudEvents is very prescriptive about the format of HTTP headers in binary mode, those two headers would have to be defined as CloudEvents extensions, because otherwise, a CloudEvent router might discard them:

```
ce-cdevents-global-id
ce-cdevents-parent-ids
```

This would be similar in a way to the existing cloudevents distributed tracing extension: https://github.com/cloudevents/spec/blob/main/cloudevents/extensions/distributed-tracing.md

I think we should still include those two values in the CDEvents context as well, so that you may have a whole JSON representation of a CDEvent that contains all the data, regardless of the binding in use.

Author: I think that's fair. I do want to highlight if you are using the HTTP protocols, the SDKs, should populate the context fields by extracting them from the headers, rather than requiring people to send duplicate information. However, we can start with duplicates, but at some point we can advance the SDKs to extracting these header values for HTTP requests.

Reviewer: @afrittoli , are you satisfied with the current wording in this PR regarding this comment now?

The reasoning for using headers instead of adding new fields to the payload, is
that some services may not be adhering to CDEvents.  It is generally easier for
a service to enable certain headers to be forwarded than needing to adapt or

> restructure some payload to accomodate some standard like CDEvents.

Reviewer 2: If I understand correctly, what you are saying is that a service may not be adopting CDEvents, but by sending these two headers we could include their messages in the sequence/tree/dag of messages.

I think that's an interesting idea, but I'm not sure it should be part of the CDEvents specification. We could write a document about how organisations may adopt CDEvents incrementally and how to cope with non-CDEvents tools (holes in the chain).

Author: Yes, that's what it is saying, but I think it is important to keep mostly cause it's indicating **why** it's a header, especially if we decide to also have it in the context. It may not be obvious.

Reviewer: @afrittoli , are you satisfied with the current wording in this PR regarding this comment now?

# Connecting Events - Links Proposal

## Abstract

This proposal will outline how to connect individual CDEvents to eachother.
Right now there's no way of associating events to one another without needing
to backtrack across certain context attributes, eg
[id](https://github.com/CDEvents/spec/blob/main/spec.md#id-context).  While
that does give us the ability to construct some graph, we do not know when a
particular span starts or finishes.

This proposal will outline a new approach that will allow for connecting
CDEvents to one another and give a clear distinction of when an activity span start and
finishes.

## Semantics

This section will line out various definitions to ensure there are no
assumptions being made when we talk about linking events

* **CI** - [Continuous integration](https://bestpractices.cd.foundation/learn/ci/)
* **CD** - [Continuous delivery](https://bestpractices.cd.foundation/learn/cd/)
* **Span** - A span is an end to end representation of all activities performed
  in the CI/CD lifecycle of some entity

> * **Link** - A link is some relation to some thing, eg event, grouping, etc.

Reviewer: As I read the Links Spec chapter below, a link is always a relation between two CDEvents.
```suggestion
* **Link** - A link is an object containing relations between two CDEvents.
```

* **Global ID** - A global ID is some overarching ID for a given span.

## Goals

The biggest challenge in this whole process is ensuring that connected events
can be retrieved quickly and efficiently, while also providing the necessary
metadata and structure to give a detailed picture of how things occurred.

1) Provide a way of quickly retrieving all related links
2) Keep link data structured small and simple
3) Scalable

## Use Cases

This section will go over a few use cases and explain how this approach can be
used to solve for each particular case.

### 1. Fan Out Fan In

The fan out fan in use case is an example where a system may make parallel
requests (fan out), and merge back to some other or the very same system (fan in)

Let us assume we have 3 system in our CI/CD environment. A continuous
integration environment, which we will call CI system, that runs tests and
builds artifacts, an artifact store that receives artifacts from the CI system,
and lastly the CD system which consume these artifacts shown in the diagram
below.

```
                
                
                
                                       

>                                       +----------+
>                                       |          |
>                     +------------+    |          |
>                 +-->| Artifact 1 |--->|          |---+       
> +----------+    |   +------------+    |          |   |    +----------+
> |          |    |   +------------+    | Artifact |   |    |          |
> |    CI    |----+-->| Artifact 2 |--->|  Store   |---+--->|    CD    |
> |          |    |   +------------+    |          |   |    |          |
> +----------+    |   +------------+    |          |   |    +----------+
>                 +-->| Artifact 3 |--->|          |---+    
>                     +------------+    |          |        
>                                       |          |
>                                       +----------+

Reviewer: I think the provided example mixes processes/flows with components and entities in a strange way. I propose to instead provide an example like the following:
```suggestion
                    +---------------+     +-------------+
                +-->| Build for Mac |---->| Test on Mac | --------+
+-----------+   |   +---------------+     +-------------+         |
|  Static   |   |   +----------------+   +--------------+         |   +---------------------------+
|  Code     |---+-->| Build for SLES |-->| Test on SLES | --------+-->| Promote & Deliver Release |
|  Analysis |   |   +----------------+   +--------------+         |   +---------------------------+
+-----------+   |   +-------------------+   +-----------------+   |
                +-->| Build for Windows |-->| Test on Windows | --+
                    +-------------------+   +-----------------+

```

Or even simpler than so, showing how a pipeline step would execute multiple test suites in parallel and then continue on to the next pipeline step in serial.

                                       
                                       
                                       

```

The diagram above shows the CI event creating 3 separate artifacts that will
make it's way to the artifact store. Some CD system would then consume those
artifact, but would need to group all the artifact when the CI system finishes
its pipeline. This is not meant to be a full diagram of the whole CDEvents
flow, but a simple representation of the artifact(s) lifecycle.

### 2. Generic UI

With the goal of interoperability, this allows for compatible standalone
generic UIs for any CI/CD system.

## Design

> This section will propose two designs. The first being how individual events

Reviewer: With this wording I'd expect two sub sections on heading level ###, but I can't find them. I think you can remove this ## Design section (5 lines) altogether

Reviewer: @xibz , I believe this comment is not yet resolved

can be linked to one another from a CDEvents perspective and define how links
will look. The second portion will address the goal of scalability.

## Connecting Events

An individual event usually has some relation to some trigger. This can be a
new commit added to a pull request, a test suite being called in the CI
process, or publishing a new artifact for some system to consume. While these
events mean something themselves, they do not give the proper context of what
caused what. This section will introduce two new fields to be used in the
CDEvent spec that will allow for giving some relation between CDEvents. The two
new fields will be added to the context section of CDEvents to solve this, a
global ID and parent IDs.

```json
{
  "context": {
    "version": "0.3.0-draft",
    "id": "505b31c2-8bc8-47b3-a1a0-269d7a8530ac",
    "source": "dev/jenkins",
    "type": "dev.cdevents.testsuite.finished.0.1.1",
    "global_id": "00000000-0000-0000-0000-000000000001", # new global id field

>     "parent_ids": ["271069a8-fc18-44f1-b38f-9d70a1695819"], # new parent id field

Reviewer: To easily cater for more types of event relations I propose to generalize the parent_ids field into a more generic `relations` field:
```suggestion
    "relations": [
      {"type": "PARENT",
       "event_id": "271069a8-fc18-44f1-b38f-9d70a1695819"},
      {"type": "ARTIFACT",
       "event_id": "1234567e-fc18-44f1-b38f-9d70a1695abc"},
     ], # new relations field
```
where ARTIFACT is just an example of a different relation, to be used for example to link a testsuite started event to an artifact packaged event.

	"global_id": "00000000-0000-0000-0000-000000000000",
	"parent_ids": ["22222222-0000-0000-0000-000000000000"],
    "source": "/event/source/123",
    "type": "dev.cdevents.pipelinerun.started.0.1.1",
    "timestamp": "2023-03-20T14:27:05.315384Z"
  },
  "subject": {
    "id": "mySubject123",
    "source": "/event/source/123",
    "type": "pipelineRun",
    "content": {
      "pipelineName": "myPipeline",
      "url": "https://www.example.com/mySubject123"
    }
  }
}
```

With these three separate CDEvents we can see that these all belong to a single
span of `00000000-0000-0000-0000-000000000000`, and that the relation is as follows

```
+---------------+   parent   +-----------------+   parent   +------------------+
| change merged |<-----------| pipeline queued |<-----------| pipeline started |
+---------------+            +-----------------+            +------------------+
             |                   ^          |                   ^
             +-------------------+          +-------------------+
              inferred connection            inferred connection
```

By establishing who has what parent, we can infer how they connect.

> ## Propogation

Reviewer: ```suggestion
## Propagation
```

used. This proposal will not cover the technical details but will cover various
use cases that can occur in large systems. A separate technical design document
can be found [here](TODO/link/to/tech/doc) for those who want some guidance on
how propagation could be handled for a given protocal

### Partial CDEvent Support

The first case is where only a subset of the systems support CDEvents.
For this example, we will use four services. Service A, B, and C.

Services A, B, and D support CDEvents, while C does not. An example CDEvent
span may look like:

```
  +-------------------------------------------------+
  |                     Event Bus                   |
  +-------------------------------------------------+
       ^           |                           ^
       |           v                           |
  +---------+ +---------+   +---------+   +---------+
  |    A    | |    B    |-->|    C    |-->|    D    |
  +---------+ +---------+   +---------+   +---------+
```

From the example above we can see that persisting a CDEvent event across A and
B can be easily done as they both support CDEvents. However, that propogation
is broken when B talks to C.

> ### Differenet Protocols

Reviewer: ```suggestion
### Different Protocols
```

As mentioned before, propogation is up to the protocol, e.g. HTTP(S), MQTT,
AMQP, etc. A full list can be found [here](https://github.com/cloudevents/spec#cloudevents-documents).
Protocols that handle propogation differently should rely on the SDKs to
consume and continue the propogation.

> ### Separate Event Busses

Reviewer: ```suggestion
### Separate Event Buses
```

> Another situation with occurs when there exists multiple event busses and

Reviewer: ```suggestion
Another situation which occurs when there exists multiple event buses and
```

know their overall flow may only contain a few links. If that is the case,
simply turning on linking payload aggregation, will send all links in the
payload. Mind you, this can make the payload very large, but may be good for
debugging.

The global ID header will continue to propogate, unless the user explicitly
starts a new CDEvent span. If there is no global ID header, the client will
generate one and that will be used for the lifetime of the whole events span

![link flow](images/links_flow.jpeg)

This show's an example of how these different types would be used in a CI/CD setting.

### CDEvents Span Definition

While most use cases can rely on the CDEvents SDK to dictate what a span may
look like. However, with the links portion of the SDK, users may specify when a
span begins and ends giving full control to the user.

### Scalability

Scalability is one of the bigger goals in this proposal and we wanted to ensure
fast lookups. This section is going to describe how the proposed links format
will be scalable and also provide tactics on how DB read/writes can be done.

The purpose of the global ID is to ensure very fast lookups no matter the

> database. We could say that only graph DBs could be used to do a full span
> lookup without a global ID but that poses two problems:
> 
> * Slower lookups as the graph DB needs to backtrack to find the full span
> * Requires either graph DBs or using SQL like graph DBs.

Reviewer: I don't fully follow this reasoning. Would it be fair to rephrase it to:
```suggestion
database. Without a global ID the database or its client would need to recursively follow event references, upstream or downstream depending on the use case. A graph DB would easily provide that, and it is also possible to implement on top of SQL like DBs and document DBs, but it will never be as fast as querying for a global ID.
```

# Connecting Events - Links Proposal

## Abstract

This proposal will outline how to connect individual CDEvents to eachother.
Right now there's no way of associating events to one another without needing
to backtrack across certain context attributes, eg
[id](https://github.com/CDEvents/spec/blob/main/spec.md#id-context).  While
that does give us the ability to construct some graph, we do not know when a
particular span starts or finishes.

This proposal will outline a new approach that will allow for connecting
CDEvents to one another and give a clear distinction of when a span start and
finishes.

## Semantics

This section will line out various definitions to ensure there are no
assumptions being made when we talk about linking events

> * **CI** - Continuous integration

Author: Disagree. Always good to be overly-thorough. One suggested change is link to the definitions instead, if that works.

> * **CI** - Continuous integration

Reviewer: It would be fine to link to existing definitions of the terms. One such target to link to could be this: https://bestpractices.cd.foundation/learn/ci/

goal of scalability.

### New Headers

To allow for connecting events and proper propogation, we will add two new HTTP
headers:
* `X-CDEVENTS-GLOBAL-ID`
* `X-CDEVENTS-PARENT-IDS`

The reasoning for using headers in addition to  adding new fields to the
payload, is that some services may not be adhering to CDEvents.  It is
generally easier for a service to enable certain headers to be forwarded than
needing to adapt or restructure some payload to accomodate some standard like
CDEvents.

```json
# Headers
# X-CDEVENTS-GLOBAL-ID: "00000000-0000-0000-0000-000000000001"
# X-CDEVENTS-PARENT-IDS: "271069a8-fc18-44f1-b38f-9d70a1695819"

{
  "context": {
    "version": "0.3.0-draft",
    "id": "505b31c2-8bc8-47b3-a1a0-269d7a8530ac",
    "source": "dev/jenkins",
    "type": "dev.cdevents.testsuite.finished.0.1.1",
    "global_id": "00000000-0000-0000-0000-000000000001", # new global id field
    "parent_id": "271069a8-fc18-44f1-b38f-9d70a1695819", # new parent id field
    "timestamp": "2023-03-20T14:27:05.315384Z"
  },
  "subject": {
    "id": "MyTestSuite",
    "source": "/tests/com/org/package",
    "type": "testSuite",
    "content": {}
  }
}
```

> Since consuming services will only have the parent IDs, it is important that

Author: It does, with the headers? I can expand it more if that's what you are asking?

> Since consuming services will only have the parent IDs, it is important that

Author: It does, by talking about the headers and propagation on line 47. I can expand it more if needed.

> Since consuming services will only have the parent IDs, it is important that

Reviewer: Ok. But the headers should probably be considered optional, and the important pieces from the spec should be the new json attributes. As discussed in other comments I believe the HTTP headers details should be moved out to the cloudevents-binding document.

the service fetches the parent whem it requires more context

## IDs

With the introduction of the two headers, it's important to establish what they
are and their respective formats.

The `X-CDEVENTS-GLOBAL-ID` is an ID that is generated when a new CDEvent span

> is wanted or if no CDEvent span is present. This ID will follow the [UUID](https://datatracker.ietf.org/doc/html/rfc4122)

Author: That's more of the technical details rather than the spec, but I can include it if needed, since this proposal does have some technical details. However, you've also mentioned to move it out of the spec proposal in one of the comments, so I guess we need to figure out what we want to include technicality wise.

> is wanted or if no CDEvent span is present. This ID will follow the [UUID](https://datatracker.ietf.org/doc/html/rfc4122)

Reviewer: You're right. Details on how to retrieve/create the global id should not be in the spec itself. But the primer could include some recommendations about it.

format.

The `X-CDEVENTS-PARENT-IDS` is any number of immediate parent IDs to satifisy
the fan out, fan in use case. These IDs will be the IDs of the `context.id` in
the CDEvent which is of the UUID format.

> ## Client APIs and Links Storage

Author: > Or do you see that it will need to be a core component in any CDEvents setup?

Not a core component, but more of a helper service like you suggested.

So far we've only talked about what a service may receive when expecting a

> parent link. However, when we store a link, there's a lot more metadata that

Author: Huh? I think the image describes it and which is why it's linked in the proposal.

I can add a more verbose section if that's what you are asking, but trying to keep a lot of the technicalities out.

> parent link. However, when we store a link, there's a lot more metadata that

Reviewer: It might be easier to understand if the picture is clarified a bit when it comes to what calls are made in what direction and in what order and such.

can and should be added.

The idea is we'd expect users to start link, group, and end links accordingly
through APIs we'd provide. This is very similar to how tracing works.
Granularity in tracing is completely up to the engineer which this proposal
also intends users to do.

This will introduce new APIs to the CDEvents SDKs, such as `addLink`.
This API will be used to create a new link based on some CDEvent context.
The context may contain things like parent caller and other useful metadata.

```
(context: CDEventContext) addLink(link)
```

startLink will utilize the parent ID to make some sort of relation back to the
parent.
This method is attached to the LinkContext which will contain the current
metadata about the current composed of link, eg what current link is being
built along with the parent link ID.

When calling `addLink`, it is important to understand the association to the
parent.

```
// Adds a new link to the CDEvent context which will be sent to the link
// service at some point.
cdEventContext.addLink(link: Link);
// Link may be a class that can be a ToLink, FromLink, WithLink
```

> Here we see an enum of `CAUSE` which is one of the few types of relations that

Author: Oh forgot to change this! Good catch!

`B`, and `C`.

Assume our pipeline was triggered by some git website, which `FooPipeline`
would create a `FromLink` indicating the trigger from our git website. Further
at some point our pipeline decides to start the `BarSuite`. `BarSuite` would
add a new `WithLink` saying it's with the `FooPipeline`. We can keep going with
other `WithLink`s for each test, but each test will be "with"-ed the suite
instead of the pipeline. It's important to note when using the `GROUP` link
type, that links that are grouped under some `ID` will be grouped **under**
that event.

Some users may prefer to not run a separate links service especially if they
know their overall flow may only contain a few links. If that is the case,
simply turning on linking payload aggregation, will send all links in the
payload. Mind you, this can make the payload very large, but may be good for
debugging.

The global ID header will continue to propogate, unless the user explicitly
starts a new CDEvent span. If there is no global ID header, the client will
generate one and that will be used for the lifetime of the whole events span

![link flow](images/links_flow.jpeg)

> This show's an example of how these different types would be used in a CI/CD setting.

Author: These are only requests.

> As I interpret it, the call from GitHub to Links Service includes a reference to the CI service ("to": "2"). Is that correct?

This may not be the source change event that you are thinking of. This is more of the CI job that run and get's started. For the source change event, would be a noop, is my guess.

So the consuming thing would create a `FROM` link is my guess

> This show's an example of how these different types would be used in a CI/CD setting.

Reviewer: Yes, please clarify that, since I still don't fully understand it :)

> This show's an example of how these different types would be used in a CI/CD setting.

Reviewer: Great! Now I understand the proposal a lot better :)

It looks unnecessarily complex to me, from a producer point of view. Why would the event producer also need to call the Link Service explicitly? Couldn't the Link Service just listen to the Event Bus and build up the links from the events seen there? I assume that all information needed in `#5` is already part of the event sent in `#4`, or?

> This show's an example of how these different types would be used in a CI/CD setting.

Reviewer: I believe this comment is now mostly sorted out, so I resolve it.

look like. However, with the links portion of the SDK, users may specify when a
span begins and ends giving full control to the user.

### Scalability

Scalability is one of the bigger goals in this proposal and we wanted to ensure
fast lookups. This section is going to describe how the proposed links format
will be scalable and also provide tactics on how DB read/writes can be done.

The purpose of the global ID is to ensure very fast lookups no matter the
database. We could say that only graph DBs could be used to do a full span
lookup without a global ID but that poses two problems:

* Slower lookups as the graph DB needs to backtrack to find the full span
* Requires either graph DBs or using SQL like graph DBs.

Instead a link service that processes and agnostically stores to some DB is
much prefer as it gives companies and developers options to choose from.  When
using an SQL database, the global ID could be the secondary key to easily
retrieve indexed entities. Links could be easily sorted by timestamp which
should roughly coordinate to their linked neighbors, parent and child.

CDEvents that are to be ingested by some service would also have to worry about

> the number of links returned. This problem is mitigated in that only the

Reviewer: It is really the `events` that are returned and not just their `links`, or have I misunderstood what a `link` is?

```suggestion
the number of events returned. This problem is mitigated in that only the
```

immediate parent(s) links are returned, and any higher ancestry are excluded.
If some service needs to get access to a higher (a parent's parent) they would

> need to use the links API to retrieve them.

Reviewer: I believe the Links Service should provide an entrypoint to be able to perform such recursive/upstream queries

Author: Definitely. I went ahead and remove all the technical details from this doc, however. I plan on following up with a tech doc to go over the technical design, whatever repo that is appropriate for

  "start_time": "2021-10-22 16:04:01.209458162 +0000 UTC",
  "end_time": "2021-10-22 16:04:01.209514132 +0000 UTC",
  "status_code": "STATUS_CODE_OK",
  "status_message": "",
  "attributes": {
    "net.transport": "IP.TCP",
    "net.peer.ip": "172.17.0.1",
    "net.peer.port": "51820",
    "net.host.ip": "10.177.2.152",
    "net.host.port": "26040",
    "http.method": "GET",
    "http.target": "/v1/sys/health",
    "http.server_name": "mortar-gateway",
    "http.route": "/v1/sys/health",
    "http.user_agent": "Consul Health Check",
    "http.scheme": "http",
    "http.host": "10.177.2.152:26040",
    "http.flavor": "1.1"
  },
  "events": [
    {
      "name": "",
      "message": "OK",
      "timestamp": "2021-10-22 16:04:01.209512872 +0000 UTC"
    }
  ]
}
```

The `attributes` field is the same as our `tags` field in this proposal. So
spans would satisfy all of our requirements, and plus some. However, there is 
complexity we will consider in a later section.

Spans also have a few different types: client, server, consumer, producer, etc.
While this can cover all our use cases, this doesn't guarantee on how libraries
like sleuth handle the span-kind for incoming requests.

> ### Context Propogation

Reviewer: If we translate this to CI/CD terms I assume we mean to connect one pipeline with another, e.g. when including a library/component into an application and connecting the events for the deliveries of that library/component into the CI/CD pipeline of the application? If so, yes that is a use case we'd need to support with CDEvents. There'd need to be a possibility to provide such cross-over event references in the events we send.

Reviewer: I believe this comment is currently out of scope for this PR

With the goal of interoperability, this allows for compatible standalone
generic UIs for any CI/CD system.

## Design

This section will propose two designs. The first being how individual events
can be linked to one another from a CDEvents perspective and define how links
will look. The second portion will address the goal of scalability.

## Connecting Events

An individual event usually has some relation to some trigger. This can be a
new commit added to a pull request, a test suite being called in the CI
process, or publishing a new artifact for some system to consume. While these
events mean something themselves, they do not give the proper context of what
caused what. This section will introduce two new fields to be used in the
CDEvent spec that will allow for giving some relation between CDEvents. The two
new fields will be added to the context section of CDEvents to solve this, a
global ID and parent IDs.

```json
{
  "context": {
    "version": "0.3.0-draft",
    "id": "505b31c2-8bc8-47b3-a1a0-269d7a8530ac",
    "source": "dev/jenkins",
    "type": "dev.cdevents.testsuite.finished.0.1.1",
    "global_id": "00000000-0000-0000-0000-000000000001", # new global id field

>     "parent_ids": [ # new parent id field
>         {
>             "link_id": "271069a8-fc18-44f1-b38f-9d70a1695819",
>         }
>     ],

Reviewer: I see the latest commit made some updates to this field, but I'd still like it to look slightly different:
```suggestion
    "relations": [ # new relations field, or should it be called "links"?
        {
            "type": "PARENT",  # or should it be called "PRECURSOR"?
            "target": "271069a8-fc18-44f1-b38f-9d70a1695819",  # the context.id of the CDEvent that this link targets
        }
    ],
```
And an example of future additions to the relations type could be "ARTIFACT", where such a link would point to the artifact.packaged (or artifact.published?) event for example.

Author: Oh whoops, I meant to to include the full cdevent type there, rather than just `ARTIFACT`.

And let's call it `PARENT`. That's what tracing uses, and I think more people are familiar with that term in computer science/programming

Reviewer: Regarding whether the "type" should be valid CDEvents event types or if they should be more generic. I can give one example, using an event type that we haven't yet added to the spec namely "confidence.reached". Such an event would be used to declare that an item has reached a certain level of confidence, often based on certain test activities being performed successfully. Such a confidence.reached event would need to somehow declare what item  has has reached the confidence. It could be for example either a source change, an artifact or a deployment. All those items (or I'd prefer to call them `subject`s) are declared using different CDEvents, and I believe we need a way to refer to them in a non-specific way. So, for a `relations` field in a "confidence.reached" event I'd like to see a definition like this:
```
"relations": [
  {"type": "parent",
   "target": "123456-123-346-234-34ade65"},  # Could point to any type of event
  {"type": "subject",
   "target": "abcdef-123-346-234-34ade32"}  # Could point to either change.created, changed.merged, artifact.published, service.deployed, and more
]
```

Other event types, like the existing "artifact.published" event, could include a relation to the "artifact.packaged" event in a similar way:
```
"relations": [
  {"type": "parent",
   "target": "abc456-123-346-234-34ade65"},  # Could point to any type of event
  {"type": "artifact",
   "target": "abc123-123-346-234-34ade32"}  # Should point to an artifact.packaged event
]
```
In this latter case, that "artifact" relations type could have been "artifact.packaged" instead, but that would make some relation types event specific and others event agnostic (like "parent" and "subject")

Author: >  I believe we need a way to refer to them in a non-specific way.

This is where I am a little confused. Why?

Reviewer: @xibz , does this reasoning make sense to you now? Do you agree that certain event types could have the same type of relation to multiple different other event types, like in the example of the (not yet proposed) confidence.reached event mentioned above where it would have a relation of type `subject` to several different upstream events?

the `subject.id` do not have a specified format, and may not be unique across
spans. For example, a CI service can have the same `subject.id` as some CD
service. Further, this field is an array which allows for events to fan out and
in like when a test suite calls a bunch of test cases in parallel.

Below represents an example of a few CDEvents and how they link to one another.

**Change Merged CDEvent**
```json
{
  "context": {
    "version": "0.4.0-draft",
    "id": "11111111-0000-0000-0000-000000000000",
    "global_id": "00000000-0000-0000-0000-000000000000",
    "parent_ids": [],
    "source": "/event/source/123",
    "type": "dev.cdevents.change.merged.0.2.0",
    "timestamp": "2023-03-20T14:27:05.315384Z"
  },
  "subject": {
    "id": "mySubject123",
    "source": "/event/source/123",
    "type": "change",
    "content": {
      "repository": {
        "id": "TestRepo/TestOrg",
        "source": "https://example.org"
      }
    }
  }
}
```

**Pipeline Queued CDEvent**
```json
{
  "context": {
    "version": "0.4.0-draft",
    "id": "22222222-0000-0000-0000-000000000000",
    "global_id": "00000000-0000-0000-0000-000000000000",

>     "parent_ids": ["11111111-0000-0000-0000-000000000000"],

Reviewer: This needs to be updated in line with the updated proposal on this field type above

    "source": "/event/source/123",
    "type": "dev.cdevents.pipelinerun.queued.0.1.1",
    "timestamp": "2023-03-20T14:27:05.315384Z"
  },
  "subject": {
    "id": "mySubject123",
    "source": "/event/source/123",
    "type": "pipelineRun",
    "content": {
      "pipelineName": "myPipeline",
      "url": "https://www.example.com/mySubject123"
    }
  }
}
```

**Pipeline Started CDEvent**
```json
{
  "context": {
    "version": "0.4.0-draft",
    "id": "33333333-0000-0000-0000-000000000000",
    "global_id": "00000000-0000-0000-0000-000000000000",

>     "parent_ids": ["22222222-0000-0000-0000-000000000000"],

Reviewer: This needs to be updated in line with the updated proposal on this field type above

Reviewer: Comment outdated

  },
  "subject": {
    "id": "mySubject123",
    "source": "/event/source/123",
    "type": "pipelineRun",
    "content": {
      "pipelineName": "myPipeline",
      "url": "https://www.example.com/mySubject123"
    }
  }
}
```

With these three separate CDEvents we can see that these all belong to a single
span of `00000000-0000-0000-0000-000000000000`, and that the relation is as follows

```
+---------------+   parent   +-----------------+   parent   +------------------+
| change merged |<-----------| pipeline queued |<-----------| pipeline started |
+---------------+            +-----------------+            +------------------+
             |                   ^          |                   ^
             +-------------------+          +-------------------+
              inferred connection            inferred connection
```

By establishing who has what parent, we can infer how they connect.

## Propagation

Span propagation will be handled differently depending on the protocol that is
used. This proposal will not cover the technical details but will cover various
use cases that can occur in large systems. A separate technical design document

> can be found [here](TODO/link/to/tech/doc) for those who want some guidance on

Reviewer: This is a dead link

Author: Right lol. It's a `TODO`

Reviewer: Either remove it or change the PR to draft, maybe. Or at least update the PR description to clearly show that this PR is a discussion base rather than something that is ready to be merged.

### Separate Event Buses

Another situation which occurs when there exists multiple event buses and
sharing occurs due to consumer(s).

```
# Case 1: Different Event Busses

+---------------+   +---------------+   +---------------+
|  Event Bus A  |   |  Event Bus B  |   |  Event Bus C  |
+---------------+   +---------------+   +---------------+
             ^ |     ^ |         ^ |     ^ | 
             | v     | v         | v     | v 
            +-----------+       +-----------+
            | Service A |       | Service B |
            +-----------+       +-----------+

# Case 2: Possible Duplicate Consumed Events

+---------------+   +---------------+
|  Event Bus A  |<->|  Event Bus B  |
+---------------+   +---------------+
             ^ |     ^ | 
             | v     | v 
            +-----------+
            | Service A |
            +-----------+
```

This proposal will not tackle these problems, but identify them as issues that
the technical design document will need to solve.

For case 1, introducing a links service becomes much more difficult. If a user
is to query for a full span since the link data may be persisted in different
databases.

> Case 2 is more about consuming duplicate events and sending duplicate events

Reviewer: I'd say that such sharing between event buses should use well-known federation principles for the bus technology in question, and the producers of events should only send events to one of the buses. If a consumer on another bus instance wants to see such an event the consumer should request the event federation to be set up accordingly.

Author: Agreed, but that isn't going to be the reality of a large company. You're going to have one siloed org with their own bus, and another siloed org with their own bus, and many years later, they are going to have some feature where they'll need to connect/talk.

Author: Yea, definitely outside the scope of this doc, but wanted to outline possible issues that we, as CDEvents, will need to think about. I can create a separate github issue for this, if needed.

Reviewer: Yes please do. I think event federation/forwarding between buses is out of scope for this PR.

Reviewer: @xibz , would you mind moving this part and instead creating a dedicated issue to sort that out?

which can cause a potential issue around how links represents those
duplications.

## Links Spec

So far we've only talked about what a service may receive when receiving the
new CDEvent containing links. However, when we store a link, there's a lot more
metadata that can and should be added.

The idea is we'd expect users to start link, group, and end links accordingly

> through APIs we'd provide. This is very similar to how tracing works.

Reviewer: I'm a bit lost in this section, and that is probably since I'm not at all fluent in "how tracing works" I guess. Would you mind referencing some valuable doc as a crash course in how tracing works here? I fail to see the need for the link_type, event_transport, and tags below which is probably due to my lack of knowledge.

Author: `event_transport` isn't needed, tbh. Tagging and `link_type` is the bare minimum. Without the link type, you wouldn't know how to group things, and where the arrows are pointing?

Reviewer: I might be naïve or have a biased view of things, but to me all "arrows" (event relations) point backwards, to some previous event. When sending a CDEvent, the producer should never know what consumer(s) will consume it.
I'm probably still misunderstanding parts of the tracing features, so you'd have to bear with me here...

Reviewer: This is kind of outdated with the changes made lately to this PR, so I'll set it to resolved.

| global_id       | CDEvent span ID                                                      |
| link_type       | An enum that represents the type of link, e.g. 'TO', 'FROM', 'GROUP' |
| event_transport | An enum that represents how the event message was transported |
| from            | Where the link is coming from                                        |
| to              | Where the link is going to                                           |
| tags            | Custom metadata that an individual link can have                     |

Below defines a list of `link_type` enums a kind can have

| Name  | Description                                                                               |
|-------|-------------------------------------------------------------------------------------------|
| TO    | When a link is creating an event, it will use TO to signal to what target ID              |
| FROM  | When a link is receiving an event, it will use FROM to signal what service called it      |
| GROUP | When a link is to be grouped with other events, it will use GROUP to establish a grouping |

> Below defines a list of `event_transport` enums a link can have

Reviewer: I believe that if the Link Service would listen to the event bus instead of expecting explicit calls from an event producer, it could determine this data by itself:
TRIGGER - If the event has no parent it means that the occurrence was triggered by something else than an event
P2P - Should CDEvents really support the use case of non-broadcasted event sending?
BUS - All events except the ones under the TRIGGER condition above?

Author: > Should CDEvents really support the use case of non-broadcasted event sending?

I mean it's documented somewhere I believe?

Author: Okay, I repurposed this to `event_kind`. It'll have enums of things  like `ARTIFACT`, `PARENT`, etc.

Reviewer: > > Should CDEvents really support the use case of non-broadcasted event sending?

> I mean it's documented somewhere in CDEvents I believe?

No, not that I know of at least. So far we've said that all CDEvents should be broadcast and not directed to a specific recipient.

Author: https://github.com/cdevents/cdevents.dev/blob/fc79b2e54502e34ca12416e01d2d7e1eaf3ee21b/content/en/docs/primer/_index.md?plain=1#L117

It does mention "why not point to point" but it doesn't say it isn't compatible? It reads more like you shouldn't do this, but you still can. If that isn't suppose to read that way, some finer words will need to be added. Also, point-to-point is in need to be supported, but that's a discussion we can have in the WG SIG

Reviewer: I believe these comments are kind of outdated by now, so I'll resolve this conversation. It might rise again if there is a need for it based on the current state of the PR.

# Connecting Events - Links Proposal

## Abstract

This proposal will outline how to connect individual CDEvents to one other.
Currently there's no way of associating events without needing to backtrack
across certain context attributes, e.g.
[id](https://github.com/CDEvents/spec/blob/main/spec.md#id-context).  While
this does give us the ability to construct some graph, we do not know when a
particular chain starts or finishes.

This proposal will outline a new approach that will allow for connecting
CDEvents and give a clear distinction of when an activity chain starts and
finishes.

## Semantics

This section will define various terms to ensure there are no assumptions being
made when we talk about linking events

* **CI**        - [Continuous integration](https://bestpractices.cd.foundation/learn/ci/)
* **CD**        - [Continuous delivery](https://bestpractices.cd.foundation/learn/cd/)

> * **Chain**      - A chain is an end to end representation of all activities performed

Reviewer 2: It may be good to have examples of chains.
NIT: those activities are not necessarily sequential - in that sense the term chain may be misleading. Since it implies a connection though, I'm happy to keep `chain`, and if we find a better name later on we can still change it.

  in the CI/CD life cycle of some entity
* **Link**      - A link is an object containing relations between two CDEvents.

## Goals

The biggest challenge in this whole process is ensuring that connected events

> can be retrieved quickly and efficiently, while also providing the necessary

Reviewer 2: In future having a reference architecture picture might help - it may not be obvious what "retrieving" here refers to - even though at least for those involved in CDEvents it will be clear

databases.

Case 2 is more about consuming duplicate events and sending duplicate events
which can cause a potential issue around how links represents those
duplications.

## Links Spec

So far we have only talked about what a service may receive when consuming a
CDEvent with links. However, when we store a link, there's a lot more metadata
that can and should be added.

The idea is we would expect users to start links, groups, and end links
accordingly through APIs we would provide. This is very similar to how tracing
works in that the individual services dictate when a tracing span starts and
finishes.  Granularity in tracing is completely up to the engineer which this
proposal also intends users to do.

This will introduce new APIs to the CDEvents SDKs to handle automatic injection
of links or allow users to provide their own definition of what a CDEvent chain
is.

The link spec definition will look like this

```json
{
  "chain_id": "05CEDE1F-8C13-4699-80C3-B4AC6468414E",
  "link_type": "RELATION",
  "timestamp": "2023-03-20T14:27:05.315384Z",
  "from": {
      "id": "F27E36A4-5C78-43C0-840A-52524DFEED03",

>       "event_kind": "ARTIFACT"

Reviewer: I believe that this `kind` should not be on either of the `from` or `to` objects, but rather on the link/relation top level, next to `link_type`. To me this is really the link_type in itself, but there might be alternative names to use since link_type is already used for another purpose.

Author: `link_kind` works as well. I just wanted to keep them a little separate since `link_type` and `link_kind` sound a little similar, which is why I opted for `event_kind`.

Reviewer: Yes, but I believe this 'kind' describes the link/relation and not the event itself, so using 'event' in that name will be misleading.

  },
  "to": {
      "id": "F004290E-5E45-45F4-B97A-FA82499F534C"
  },
  "tags": [
    "ci.environment": "prod"
  ]
}

> ```

Reviewer: It is now proposed to also add an optional list within the context object of the CDEvents spec to hold this same information, as an alternative to use the standalone links spec. For that I propose a syntax like this for that list:

```json
[{
  "link_type": "RELATION",
  "link_kind": "ARTIFACT"
  "from": "F27E36A4-5C78-43C0-840A-52524DFEED03",
  "tags": [
    "ci.environment": "prod"
  ]
},
...]
```

The `chain_id` and `timestamp` are already part of the CDEvents context, so they don't need to be repeated in this list of links objects.

`to` is implicit, since it is the id of the current event

I'm not too fond of the name `from` to describe the other event in this relation/link. To me that other event is rather the `target` of this relation/link, so `target` would make more sense to me.

I moved the `event_kind` to the top level of the object, and renamed it to `link_kind` to better match what it would describe to me. The name of the parameter is not a perfect choice maybe.

When including the link within the event itself, some enums of the `link_type` becomes superfluous to me. `START` is implicit when there is no link at all. `RELATION` is what all links are about I think, but I might not see the difference to `GROUP` here. `END` kind of makes sense to me, but a link that is there to mark the end of a chain could at the same time have a `RELATION` to some other event I think. So it would be clearer to me to include a boolean parameter called `end_of_chain` or similar.

In this structure, the `from` (or `target`) parameter is maybe the only mandatory one. The `link_kind` is kind of mandatory in my view, to understand what kind of relation this link describes. Maybe `link_type` could be mandatory as well, depending on how my other comments on that parameter are answered.

Author: Yea, I think this works well. I choose `to` and `from` cause it was simple. `target` would be `to` and not `from` though, right?

So Im thinking
`to` -> `target`
`from` -> `source` or `origin`?

The rest of your comments I agree with, but Im also okay with `to` and `from`. I feel like it's super simple and conveys the direction very well.

Reviewer: My concern is mostly based on what it would look like when we include the link in the event itself. The 'current' event id will then not be needed, so the only event id needed in such a link is the 'other' event. I'd like to call that 'other' event the `target`, since the link emanates from the 'current' event and points backwards in time as I see it. Do you see that the link points forwards or backwards in time?

In an event-driven/descriptive flow, the earlier event will not know what later event(s) to link to, but the later event(s) will know what earlier event(s) to link to, so to me it is natural to name the earlier event as the `target` of the link.

Author: > Do you see that the link points forwards or backwards in time?
Links are connected backwards in time, but are a forward directional DAG with respect to time.

Maybe the `START` can just be no links, and a `chain_id`. That may work. The issue is what if it's a single node. You can't display that unless you explicitly message start which can be done either with the same way of sending links with a `START` or infer it based on no links and a `chain_id`, but then what if this is an island in a DAG instead of a `START`? Without `START` you have to make the assumption that any event with no links is a start or potentially a starting node.

Reviewer: Not sure I'm following fully. The process that sends the first event in a chain would send an event with a chain_id and no links. Could maybe be a change.created event for example. That process will not necessarily know that any other process will pick it up and start running a pipeline for it. If there is then that other process would reuse the same chain_id I suppose, and set a link to target that change.created event in its first event sent. But before that happens the change.created event will be an island event with just a chain id and no links. Is there a problem with that?

Author: You basically said what I was trying to say, and I was wondering if that is how we wanted to work is all. Sounds like it is how we want this to work :)

Reviewer: So do you think my suggestion in the first comment in this thread is more or less valid to bring in to the event schemas? I don't mind if some of the fields have slightly different names, but we should definitely use the same names on the link object parameters regardless of if they are given within the CDEvents protocol or as a separate links event protocol.

Author: Yea, Im still debating target. I've been updating the schemas to include these optional links now, and I think `target` is a little confusing, since it doesn't imply direction, people may assume forward in time instead of backwards. Opposed to `to` and `from` which is clear in which direction. So if we do choose a word, I think it needs to have some direction to it

Reviewer: I totally agree that the links need to have a clear direction. There is no way to say, in an event-driven/descriptive setup, what event(s) will be sent later, so the only option as I see it is to link backwards in time, to events that have happened earlier. The chain of linked events should form a DAG, to make it a deterministic and non-recursive operation to lay them out in the correct order.

When the links are sent separately there is of course a possibility to set the direction either way, since both the earlier and the later events have then been sent, but when including the link in the event itself there is no way to declare the link in the earlier event.

Regarding the name, to me `target` says what the link points `to` and that would be the earlier event.

Author: Wouldn't `from` or `to` work then, rather than target? When it is pointing backwards it uses `from` and when it wants to point forward it uses `to`?

As opposed to `target` which can be either direction. How would you know it's pointing forward or backwards?

And regarding connecting to other chain ids, Im not sure I understand why you would want to connect different chains. Let's leave it out for now, and if the use cases pop up, we can always amend the spec to include it

Reviewer: My thinking is that a link always points backwards. If it is well known that that how links are directed, then I think `target` could be valid.
My concern against `from` is related to the intended use of `link_kind` (or what we'll call it), that describes what kind of relation the events have to each other. It must be the producer of the later event that declares that relation, since the producer of the earlier event will not know what events a listener might send and what it will be used for.

For example:
A build tool could be the producer of an artifact.packaged/published event for a built Docker image for example. The artifact.published event might be listened to by many different consumers, i.e. a deployer, an archiver, a binary scan service, a test service, etc. All those consumers could send their own events and they'd like to link their new events to the artifact.published event. A deployer might set the `link_kind` to `IMAGE`. A binary scan service might set the `link_kind` to `BINARY`. A test service might set the `link_kind` to `IUT`. And so on. The build tool itself could also add a link between the new artifact.published event to the previous artifact.published event sent for the same type of build, and by that setting the `link_kind` to `PREVIOUS` for example.
In other words, the name/type/kind of the relation should be given in the context of the producer of the event that needs to link to an earlier event.

Author: Thanks for the example. This reminds me of the old group link event I had but removed since I assumed the new field would handle it better. This is what it solved in that it allowed to just say it was apart of something. Relations are meant to be a directional connection, but in this case it's being used as a way to group an artifact to some event or chain. Does it make sense to bring back group to handle exactly what you are talking about?

So the 4 events:
Start: indicates the beginning of a chain 
End: indicates when a chain has completed 
Relation: a directional relation to another event 
Group: a grouped relation to another event or chain

The `link_kind` enum can just be moved to group link events instead.

Author: Okay, so before I push up these changes, I just want to ensure the changes I am making make sense.

`RELATION` with `from` and `to` renamed to `PATH`
`RELATION` is now more closely related to what I envisioned `GROUP` to do.

```json
{
  "chain_id": "D0BE0005-CCA7-4175-8FE3-F64D2F27BC01",
  "link_type": "PATH",
  "timestamp": "2023-03-20T14:27:05.315384Z",
  "from": {
      "id": "38A09112-A1AB-4C26-94C4-EDFC234EF631" # context.id of #1
  },
  "to": {
      "id": "AA6945F8-B0F1-48DD-B658-25ACF95BD2F5" # context.id of #5
  },
  "tags": {
    "ci.environment": "prod"
  }
}
```

Notice the three major changes. The first is the link with `from` and `to` is a `PATH` type. Further I made `from` and `to` objects. This will allow for extra metadata that may be needed to connect things in the future, like part of a different chain?

Then for the updated `RELATION`

```json
{
  "chain_id": "8b409a3b-a706-490f-a06b-e0aea7c1028d",
  "link_type": "RELATION",
  "link_kind": "ARTIFACT",
  "timestamp": "2023-03-20T14:27:05.315384Z",
  "target": "8ae4eef5-4ecb-4cbd-9d31-4120afe0a2b1",
  "tags": { "bec2f2b2-f4b9-4ec1-bebe-4baf20ac58a3",
    "ci.environment": "prod"
  }
}
```

Now, for embedded like we've been talking about, it'd be all that we talked about except when you want to relate things, you can with a separate link type

Reviewer: Now you make me puzzled. You've renamed `to` and `from` to `source` and `target`. That might be ok. I'm not sure what happened to `link_ type` though. In the first example above it is `PATH` and in the second it is `RELATION`. And the event id parameters/objects are named differently for those different `link_types`? So, is `PATH` what was previously known as `RELATION` and `RELATION` has now go a new meaning, more like how you defined `GROUP` before? And when the `link_type` is set to `RELATION` there is a `link_kind`, but when `link_type` is `PATH` there is no `link_kind`? Why? Is the `PATH` to describe the causal relationship and `RELATION` to describe all other relationships? If so, why can't they all be `RELATION` and the `link_type` could be `CAUSE` when the relationship is a causal one?

And, what would these examples look like if they where inline in the event itself?

| chain_id   | This represents the full lifecycles of a series of events in CDEvents            |
| link_type  | An enum that represents the type of link, e.g. 'START', 'END', 'RELATION'        |
| event_kind | A stringed value representing any sort of relationship the link has to the event |
| from       | Where the link is coming from                                                    |
| to         | Where the link is going to                                                       |
| tags       | Custom metadata that an individual link can have                                 |

Below defines a list of `link_type` enums a kind can have

| Name     | Description                                                                               |
|----------|-------------------------------------------------------------------------------------------|
| START    | The starting CDEvent in a chain                                                           |
| END      | The ending CDEvent in a chain                                                             |
| RELATION | When a link has a relation, it can describe that relation with the `to` and `from` fields |

> | GROUP    | When a link is to be grouped with other events, it will use GROUP to establish a grouping |

Reviewer: Some more explanation on how this differs from a `RELATION` would be good, for clarity

Author: `GROUP` was before `link_kind`, so it made it easier to understand how things were grouped together.  However, now that we have `link_kind` this can now go away, cause now you can just have `ARTIFACT` for example. Which allows for querying just for that specifically. As this was trying to do the same thing but with utilizing `GROUP` and tags.

works in that the individual services dictate when a tracing span starts and
finishes.  Granularity in tracing is completely up to the engineer which this
proposal also intends users to do.

This will introduce new APIs to the CDEvents SDKs to handle automatic injection
of links or allow users to provide their own definition of what a CDEvent chain
is.

The link spec definition will look like this

```json
{
  "chain_id": "05CEDE1F-8C13-4699-80C3-B4AC6468414E",
  "link_type": "RELATION",
  "timestamp": "2023-03-20T14:27:05.315384Z",
  "from": {
      "id": "F27E36A4-5C78-43C0-840A-52524DFEED03",
      "event_kind": "ARTIFACT"
  },
  "to": {
      "id": "F004290E-5E45-45F4-B97A-FA82499F534C"
  },
  "tags": [
    "ci.environment": "prod"
  ]
}
```

| Name            | Description                                                        |
|-----------------|--------------------------------------------------------------------|
| chain_id   | This represents the full lifecycles of a series of events in CDEvents   |
| link_type  | An enum that represents the type of link, e.g. 'START', 'END', 'RELATION', 'GROUP'    |

> | event_kind | An enum that represents a category of what the event is associated with |

Reviewer 2: We already have event types and event "buckets" - I'm not convinced we need an `event_kind ` here

Author: @e-backmark-ericsson Since you were suggesting this field. Not exactly the names or all the enums, but something that had some context to the actual subject. What is the reasoning here? Cause I do agree with @afrittoli, but added it cause it didn't seem to hurt.

Reviewer: Well, I don't think I've proposed the term `event_kind`, but what I was after was a way to declare what kind of relationship the `from` event has to the `to` event. For example. a service.deployed event would be related to an artifact.published event through an `ARTIFACT` relation, an artifact.packaged event would be related to a change.merged event through a `SOURCE` relation and a testSuiteRun.queued event would be related to a service.deployed event through an `IUT` relation.

| from       | Where the link is coming from                                           |
| to         | Where the link is going to                                              |
| tags       | Custom metadata that an individual link can have                        |

Below defines a list of `link_type` enums a kind can have

| Name     | Description                                                                               |
|----------|-------------------------------------------------------------------------------------------|
| START    | The starting CDEvent in a chain                                                           |
| END      | The ending CDEvent in a chain                                                             |
| RELATION | When a link has a relation, it can describe that relation with the `to` and `from` fields |
| GROUP    | When a link is to be grouped with other events, it will use GROUP to establish a grouping |

> Below defines a list of `event_kind` enums a link can have

Reviewer: The names chosen here, at least "CD" and "CI" could be interpreted quite broadly. I assume that ARTIFACT intends to be used to relate any event to an artifact.packaged event, and BUILD intends to be used to relate any event to a build.queued (or started/finished?) event. But what is "TEST" intended to be used for? Could I expect a testcaseRun or a testsuiteRun or some other event there? What could the "CI" and "CD" types be used for?
I believe that we should either have a free-text field here, giving the possibility for the producer to name the link type anyway they want, or we should have dedicated link types tied to specific event types, so that the consumer could be certain about what to find on the "source side" of the relation.

Author: These enums are more about how users want to categorize things. I can definitely see any of the test CDEvents being labeled this way with `TEST`, and even a testing environment could be labeled with this.

And the `CI` and `CD` were meant to be broad. Basically, if you still wanted to categorize something under the `CI` system, you can.

Author: Okay, that works for me.

The biggest thing I thought about is that if an organization adds a new service from open source or dependency that supports CDEvents which later get rendered as a nice graph by utilizing CDEvents, then that context is missing for any new dependency or services that they pull in and will require code changes to have it support their format.

I think the services or dependencies can maybe make this configurable, but not too sure what that looks like.

| Name     | Description         |
|----------|---------------------|
| ARTIFACT | An artifact CDEvent |
| BUILD    | An artifact CDEvent |
| CD       | A CD event          |
| CI       | A CI event          |
| TEST     | A test CDEvent      |

These links can be, but are not limited to being, sent when a CDEvent has
completed: to some collector, the link service, or the event bus . Further
the link service will allow for tagging of various metadata allowing for users
to specify certain labels when looking at links.

Some users may prefer to not run a separate links service especially if they
know their overall flow may only contain a few links. If that is the case,
simply turning on link payload aggregation, will send all links in the
payload. Mind you, this can make the payload very large, but may be good for
debugging.

The chain ID header will continue to propagate, unless the user explicitly
starts a new CDEvent chain. If there is no chain ID, the client will generate
one and that will be used for the lifetime of the whole events chain

Further, links are always constructed from the consumer point of view, with the

> exception being the start and stop links. Allowing for start and stop to be

Author: Right. This is talking about when sending the events and how to link events that aren't attached. The example you listed is a case where every entity knows how things are attached, so they can create the link. I'll make that more clear.

   new chain.

```json
{
  "chain_id": "D0BE0005-CCA7-4175-8FE3-F64D2F27BC01",
  "link_type": "START",
  "timestamp": "2023-03-20T14:27:05.315384Z",
  "id": "38A09112-A1AB-4C26-94C4-EDFC234EF631" # context.id of #1
}
```

3. Event bus proxies the link payload from `#2` to the links service.

4. Shows some consumer consuming `#1` to do some action.

5. The CI system will queue a pipeline execution, and will generates a
`context.id` to be sent

```json
{
  "context": {
    "version": "0.4.0-draft",
    "chain_id": "D0BE0005-CCA7-4175-8FE3-F64D2F27BC01",
    "id": "AA6945F8-B0F1-48DD-B658-25ACF95BD2F5",
    "source": "/event/source/123",
    "type": "dev.cdevents.pipelinerun.queued.0.1.1",
    "timestamp": "2023-03-20T14:27:05.315384Z"
  },
  "subject": {
    "id": "mySubject123",
    "source": "/event/source/123",
    "type": "pipelineRun",
    "content": {
      "pipelineName": "myPipeline",
      "url": "https://www.example.com/mySubject123"
    }
  }
}
```

6. As the send change merge event is sent, the system will follow up with
sending a link associated with the prior event which connects `#1` to `#5`

```json
{
  "chain_id": "D0BE0005-CCA7-4175-8FE3-F64D2F27BC01",

>   "link_type": "RELATION",

Reviewer: So, this link instantiates the relation between the pipelinerun.queued event and the change.merged event. To me the relationship (to/from) is in the wrong order. I'd like to be able to say that the pipelinerun.queued event has a relationship _to_ the change.merged event, and not the other way around. Furthermore I'd like to be able to describe what kind of relation this is. In this case I'd say it's a causality relation, i.e. the change.merged CAUSEd the pipelinerun to be queued, so the CAUSE of the pipelinerun.queued is the change.merged event.
And to me, it would be very natural to include that relationship info as part of the pipelinerun.queued event itself, instead of having to create a separate event for it. Sure, and SDK could simplify that to its user, so only one call to the SDK would be needed, but it will be harder to troubleshoot without being dependent on dedicated tools that can combine the data from the CDEvent with the links event.

Reviewer: Can't find this example anymore, so I assume it has been corrected

  "chain_id": "D0BE0005-CCA7-4175-8FE3-F64D2F27BC01",
  "link_type": "END",
  "timestamp": "2023-03-20T14:27:05.315384Z",
  "id": "7D5E011F-5073-44A7-B4F0-86DD7D4C2C7F" # context.id of #31
}
```

### Separate Links vs Within CDEvents

This will go over the design choice of having links as a separate event
rather than being contained within some CDEvent. The most important design
consideration is that links could expand the payload of a CDEvent especially
for a large complex systems. A system may have a deployment of hundreds of
micro-services as artifacts to kubernetes. If we consider having the links live
within the CDEvent, then this specific deployment would contain links to all
consumed artifact events as parents, which greatly expands the total size of
the CDEvent. By having them as separate events allows users to more easily
consume CDEvents. Further, a counter argument may be that those hundred
artifacts could be represented as a single artifact. While that is true, we are
forcing users to know that they need to compress all those artifacts into a
single artifact, because we decided to include links within the CDEvent itself,
and this also restricts how users use CDEvents.

> Further links act as metadata about how events are connected, and do not

Reviewer: Links isn't just about relating how events are connected, even though that might be the most important use case. An almost as important use case is to decrease the payload of an event by having links referencing other events that contain the data instead of duplicating the data into each event. For example, instead of including all data about an artifact in each testcaseRun event, we instead link to the artifact.packaged event from each testcaseRun event.

Author: > An almost as important use case is to decrease the payload of an event by having links 

This is not a use case. This is just what the links are. Links are just meant to give metadata on how events are connected.

> For example, instead of including all data about an artifact in each testcaseRun event, we instead link to the artifact.packaged event from each testcaseRun event.

We've talked about this before, and I also call that out explicitly in this proposal. Basically the summary is that people should be able to use CDEvents however they see fit. There are a few other reasons to why we don't want to do that, but I think just the single reason of allowing users to use CDEvents however is necessary.

Reviewer: Well, that's true as long as the organization(s) are aligned on what to expect from the events. The more freedom the spec gives to their implementations, the more alignment operations are needed in the organizations involved. The edge case of this would of course be to use custom events everywhere, and the other edge case would be to provide no freedom whatsoever in the spec or the use of it. The more the protocol dictates what the producers can create, the easier it will be for the consumers to understand the content.

I believe we need to find a good common ground, and that's what the CDEvents spec seeks to do. The things we cannot align on (at this moment) we should leave customizable, and the things we can align on we should specify.

I also believe that the protocol should dictate _the_ way of expressing something, and not provide multiple ways to express the same thing. I.e., if you've created an artifact and want to notify about that, you should use the artifact.packaged event and include all mandatory fields in there. The more variants of CDEvents are implemented, the harder it will be for its consumers to make sense of the data sent.

Author: > I also believe that the protocol should dictate the way of expressing something, and not provide multiple ways to express the same thing

> if you've created an artifact and want to notify about that, you should use the artifact.packaged event and include all mandatory fields in there. The more variants of CDEvents are implemented, the harder it will be for its consumers to make sense of the data sent.

So there's already multiple ways of doing something based off these two comments. Unless the specification somehow dictates or mentions that, we cannot enforce that. This gets into the world of hurt when you start considering open source projects who don't support this notion. 

> artifact.packaged

This is confusing. Why would someone listen to an artifact packaged event for this? The name is already confusing if that is the intent.

Like even the [description](https://github.com/cdevents/spec/blob/v0.3.0/continuous-integration.md#artifact-packaged) of what it is seems like the wrong thing to listen for and `artifact.published` seems more appropriate.

Reviewer: Well, the artfact.packaged event was just an example. I agree that most cases it would make more sense to listen for an artifact.published event. My intent with that comment was that if you want to notify a consumer of something that has happened in your CICD system, then there should be _one_ way to do so. And to me, the link relating e.g. an artifact.packaged event with an artifact.published event is also something that should be done in _one_ way only. If there are multiple ways to relate events to each other then the consumers need to handle all those cases to be compliant with the spec.

I'd like to see the links/relations being defined as part of the CDEvents event spec for that reason, that there should be one way to declare not just the event data but also their relations.

I don't see the issue with the packaging of 1000 artifacts. If the producer considers them 1000 different artifacts, then it would send 1000 artifact.packaged events. If the producer instead would like to notify about the 1000 artifacts as one super-artifact and send just one artifact.packaged event, that's perfectly fine as well.

Author: > I don't see the issue with the packaging of 1000 artifacts. If the producer considers them 1000 different artifacts, then it would send 1000 artifact.packaged events. If the producer instead would like to notify about the 1000 artifacts as one super-artifact and send just one artifact.packaged event, that's perfectly fine as well.

> I'm sorry, but I fail to follow your reasoning on what issue is mitigated by sending separate events for links.

This makes the links of the CDEvent for the next event 1000 links That's the issue here. Is the fan out of 1000, to be consumed by some CD system, which becomes the fan in. So, that's primarily my concern is that for system with a lot of artifacts, tests, etc, could have a lot of links if users do not use the appropriate event (which third party libraries, frameworks, or services may or may not).

This is strictly looking at the fan out and fan in situation

Author: 1000 artifacts or 1000 test cases/suites. Like if CI passed and had 1000 test cases, the pipeline finished event would have all the links to all test cases or test suites. For artifacts, our CI system for our services creates about 500 artifacts, so when pipeline finishes, links to all 500 of the artifact would be embedded unless they explicitly use the artifact packaged event. 

So, to give some context, with tracing we had a team send a span for every single possible atomic operation. This flooded the tracing graph and made it quite literally a pain in the ass to view. Tracing has "what you should do" and then there is what people may do. So I've never been a fan of something that says "oh you should do it this way", with no way of enforcing it. Im a bit of a cynic in this context.

Reviewer: Ok. In our case I have never seen a CI pipeline producing 1000 artifacts from one single source change, but several thousand test cases are for sure executed on a single source change. Those test cases are often grouped into test suites though, and the test case events would only be linked/related to the test suite they are executed within and not to all of the suites in the CI/CD pipeline. When we then fan-in later in the pipeline we would link/relate to those test suite executions rather than the individual test case executions. So all-in-all there could be several thousand events in one event chain, but they are not all related explicitly to one specific event, but rather interlinked between each other forming a DAG where each link/relation also has a link type (see my comment on the `event_kind` field in your proposal) to make it easy to look up the wanted events out of that large graph.

Author: Understood. But again you are relying on what people (dependencies) to be sending. What you and I believe is the best way to send something is meaningless if it isnt enforced anywhere; which it isnt. So, to alleviate the occurence of some dependency and/or service, this allows for easy mitigation, rather than needing to wait for the dependency needing to be updated, refactored, etc. It's much more flexible this way

Author: > but for the scenarios where the amount of links from one event is very low
Yes, but we aren't looking to solve only a subset of the problems, especially for companies like ours, where this number of artifacts is really normal. Further we have 1000s upon 1000s of tests that are ran for a single job all across different teams here at Apple. So while it might work for other companies, for companies like ours, and my experience at AWS, this is going to be an issue if we don't have them separated.

And Im still very confused on why the debate on this has lasted so long, tbh. Like we have issues with this very thing. Is that not enough to at least implement it this way, and if we need to update the spec, we can always embed them and have that be configurable. I understand you want one way, but if we choose one way, embedding them is not an option for us.

Reviewer: The original comment here seems not to be relevant anymore since the recent updates of this PR

    }
  }
}
```

## Propagation

Chain propagation will be handled differently depending on the protocol that is
used. At a high level, the SDK expects calls to construct the proper paths, but
connecting of events will be handled completely by the SDKs.

There are a couple cases where propagation is difficult, and maybe even
impossible depending on what the receiving services are willing to do. In
events where the SDKs are not used, it is up to these receiving services to
pass the chain IDs. If this is not done, then there will be a break in the
chain, which means that a new chain will be started, or missing links that will
cause these services to not be visible.

## Links Spec

So far we have only talked about what a service may receive when consuming a
CDEvent with links. However, when we store a link, there's a lot more metadata
that can and should be added.

The idea is we would expect users to start links, connect links, and end links
accordingly through APIs we would provide. This is very similar to how tracing
works in that the individual services dictate when a tracing span starts and
finishes.  Granularity in tracing is completely up to the engineer which this

> proposal also intends users to do.

Author: Oh, they can absolutely not send an `END` chain. However, the `END` chain is primarily useful for systems that want to display some lifecycle from point A to B.

Without `END` chain types, it just becomes a little bit more difficult in deciding when a system is continuing versus not, especially when we are talking about allowing any tool to be able to communicate with CDEvents. The frontend may have no idea where the starting point is and it wont know where the end point is either. So being a little more explicit allows for letting the frontend know when something is "done" (whatever that means). Right now it isn't obvious what is going to send finish, but that could be up to the company in deciding who and what sends it.

> proposal also intends users to do.

Reviewer 2: I think that dashboard and other systems collecting events will have to be flexible enough to no rely on an end-message. In distributed tracing there is no end message, because it's very unlikely that a system has the context to know that it's an end message, unless, like you mentioned, a company sets that explicitly.
In any case, I'm ok to keep the END message in, but in future we might add some further context to the docs about them.

This will introduce new APIs to the CDEvents SDKs to handle automatic injection
of links or allow users to provide their own definition of what a CDEvent chain
is.

These links can be, but are not limited to being, sent when a CDEvent has
completed: to some collector, the link service, or the event bus. Further
the link service will allow for tagging of various metadata allowing for users
to specify certain labels when looking at links.

Some users may prefer to not run a separate links service especially if they
know their overall flow may only contain a few links. If that is the case,
simply turning on link payload aggregation, will send all links in the
payload. Mind you, this can make the payload very large, but may be good for
debugging.

The chain ID header will continue to propagate, unless the user explicitly
starts a new CDEvent chain. If there is no chain ID, the client will generate

> one and that will be used for the lifetime of the whole events chain.

Author: fan out would have one chain id?

When we say fan out I imagine, one request branching out into many. Now if you are talking about fanning multiple requests into one, then that may be a little bit more tricky, and may have to create a new chain, with relation links

> one and that will be used for the lifetime of the whole events chain.

Reviewer: Maybe we would need to change the chain_id into an array to cope for that case? However, the chain_id is optional, and you could very well build up your linked graph of events without it. From my perspective it's mostly there for convenience and performance reasons, to quickly fetch all events from the same chain without needing to traverse a graph to get them all.

> one and that will be used for the lifetime of the whole events chain.

Author: I don't think it needs to be an array, but yea, worst case we can always change it, but I believe this would be handle by relation links.

> one and that will be used for the lifetime of the whole events chain.

Reviewer 2: ok, so for now we only support one chain_id i.e. fan-out / fan-in and if needed we can add later support for fan-in of multiple workflow.

  },
  "subject": {
    "id": "mySubject123",
    "source": "/event/source/123",
    "type": "change",
    "content": {
      "repository": {
        "id": "cdevents/service",
        "source": "https://github.com/cdevents/service/pull/1234"
      }
    }
  }
}
```

Something to call out here is that the `chain_id` may have been `null`, for
whatever reason, prior to this event. This means any parents to this event did
not generate a `chain_id`. When an event is sent, it is important that the
sender generates this id.

2. We send the start link to let the links service know that we are creating a
   new chain.

```json
{
  "chain_id": "d0be0005-cca7-4175-8fe3-f64d2f27bc01",
  "link_type": "START",
  "timestamp": "2023-03-20T14:27:05.315384Z",
  "start": {
    "context_id": "38a09112-a1ab-4c26-94c4-edfc234ef631" # context.id of #1
  }
}
```

3. Event bus proxies the link payload from `#2` to the links service.

4. Shows some consumer consuming `#1` to do some action.

5. The CI system will queue a pipeline execution, and will generates a
`context.id` to be sent

```json
{
  "context": {
    "version": "0.4.0-draft",
    "chain_id": "D0BE0005-CCA7-4175-8FE3-F64D2F27BC01",

>     "id": "AA6945F8-B0F1-48DD-B658-25ACF95BD2F5",

Reviewer: lowercase ids

    "source": "/event/source/123",
    "type": "dev.cdevents.pipelinerun.queued.0.1.1",
    "timestamp": "2023-03-20T14:27:05.315384Z"
  },
  "subject": {
    "id": "mySubject123",
    "source": "/event/source/123",
    "type": "pipelineRun",
    "content": {
      "pipelineName": "myPipeline",
      "url": "https://www.example.com/mySubject123"
    }
  }
}
```

6. As the send change merge event is sent, the system will follow up with
sending a link associated with the prior event which connects `#1` to `#5`

```json
{
  "chain_id": "d0be0005-cca7-4175-8fe3-f64d2f27bc01",
  "link_type": "PATH",
  "timestamp": "2023-03-20T14:27:05.315384Z",
  "from": {
      "context_id": "38a09112-a1ab-4c26-94c4-edfc234ef631" # context.id of #1
  },
  "to": {
      "context_id": "aa6945f8-b0f1-48dd-b658-25acf95bd2f5" # context.id of #5
  },
  "tags": {
    "ci.environment": "prod"
  }
}
```

7. The event bus will then forward link `#6` to the links service

34. Paylods from `8-33` are very similar to all prior payloads shown here, but
    the last sequence is the ending link.

```json
{

>   "chain_id": "D0BE0005-CCA7-4175-8FE3-F64D2F27BC01",

Reviewer: lowercase id

#### Start Link

Start links are used to indicate that a new chain has been started, and is a
special type of `PATH` link. The reasoning for having a separate link type for
both `START` and `END` is that it allows for clear indication of starting and
stopping a chain. If we relied only on `PATH` link types to indicate either of
these states, then consumers may not be able to distinguish the two different
states.  This makes it very clear and easy for consuming systems.

| Name  | Description                                                                        |
|-------|------------------------------------------------------------------------------------|
| start | An node object that describes the event that is associated with starting the chain |

```json
{
  "chain_id": "97ef590e-0285-45ad-98bb-9660ffaa567e",
  "link_type": "START",
  "timestamp": "2023-03-20T14:27:05.315384Z",
  "start": {
    "context_id": "a721d6ba-bbd6-4737-9274-5ddd2526b92f"
  },
  "tags": {
    "ci.environment": "prod"
  }
}
```

#### End Link

End links are used to indicate that a new chain has completed, and are a

> special type of `PATH` link.

Reviewer 2: When is an `END` link used, and what benefit does it bring?
If CDEvents is adopted correctly, any consumer in the chain will also be a producer. 

```
A (PR) -> B (Build) -> C (Test) -> D (Deploy)
```

In the chain above `D` will receive an event, and will perform a deployment. 
`D` should send the `END` link event, as well as a `service.updated` CDEvent. 
But how may `D` be sure that no-one else will pick up the `service.updated` event and continue the chain?

If the structure of the chain is known in each element contributing to the chain, then adding an element to the chain would require updating all other elements, nullifying the flexibility provided by the event-based architecture.

Author: I am going to resolve this one as it is kind of a duplicate to the response and question above. And I believe I answered this in the other comment as well

  "timestamp": "2023-03-20T14:27:05.315384Z",
  "source": {
    "subject_id": "5668c352-dd9d-4dee-b334-384e4661d21b"
  },
  "target": {
    "context_id": "3579a5aa-ef46-4ee8-95db-0540298835de"
  },
  "tags": {
    "ci.environment": "prod"
  }
}
```

The `source` field allows for either providing the either the `subject_id` or
the `context_id`. This allows for relation of different ID types provided by
CDEvents while also being explicit in what the `id` is.

### Scalability

Scalability is one of the bigger goals in this proposal and we wanted to ensure
fast lookups. This section is going to describe how the proposed links format
will be scalable and also provide tactics on how DB read/writes can be done.

The purpose of the chain ID is to ensure very fast lookups no matter the
database. Without a chain ID the database or its client would need to
recursively follow event references, upstream or downstream depending on the
use case. A graph DB would easily provide that, and it is also possible to
implement on top of SQL like DBs and document DBs, but it will never be as fast
as querying for a chain ID.

Instead a link service that processes and stores the links to some DB is much

> prefer as it gives companies and developers options to choose from.  When using

Reviewer 2: NIT: preferable / preferred

works in that the individual services dictate when a tracing span starts and
finishes.  Granularity in tracing is completely up to the engineer which this
proposal also intends users to do.

This will introduce new APIs to the CDEvents SDKs to handle automatic injection
of links or allow users to provide their own definition of what a CDEvent chain
is.

These links can be, but are not limited to being, sent when a CDEvent has
completed: to some collector, the link service, or the event bus. Further
the link service will allow for tagging of various metadata allowing for users
to specify certain labels when looking at links.

Some users may prefer to not run a separate links service especially if they
know their overall flow may only contain a few links. If that is the case,
simply turning on link payload aggregation, will send all links in the
payload. Mind you, this can make the payload very large, but may be good for
debugging.

The chain ID header will continue to propagate, unless the user explicitly
starts a new CDEvent chain. If there is no chain ID, the client will generate
one and that will be used for the lifetime of the whole events chain.

> Further, when links are created from a producer or consumer, the producer will
> not know what is consuming a CDEvent. This makes connecting CDEvents to one
> another difficult from the producer. To mitigate this, links will be
> constructed from the consumer with the exceptions of the start and stop links.
> When a service knows how CDEvents are connected, the service can choose to
> connect them themselves. A good example of this is when a service runs tests.
> There is no need for the consumer to construct this, and can all be done from
> the service running the test. 

Reviewer: I find this paragraph quite hard to read still. And from our latest discussion on this I understood that we might have different views on what a 'producer' and 'consumer' is. Would this suggestion resonate with you?

```suggestion
In the case when an event consumer is also an event producer, such a consumer
will easily be able to construct the link to the consumed event. A producer of events
with start links is often not an event consumer itself. When links are sent stand-alone,
and not embedded within the CDEvents themselves, an event consumer that is not
an event producer could still construct the links if it has the necessary information for
it. An event producer that wants to provide a link to some earlier sent event will need
to look it up in a links service or similar if the producer wants to embed the link into
the produced event itself. For events produced within a service, that service should
be able to construct the links between those events by itself.
```

wanted or if no CDEvent chain is present. This ID will follow the
[UUID](https://datatracker.ietf.org/doc/html/rfc4122) format. Chain IDs will
serve as a bucket for all CDEvents with some sort of path to each other.

### Optional Links Field

While links can be sent separately, links can also be embedded in the CDEvent
context as an optional field. Embedded links look similar to separate links
except `START` links are not needed since we can infer when a chain has started
based on the context of the event. This leaves the only two types of links that
may be embedded, `PATH` and `END`. While `END` could be also inferred it
may is not as simple especially if there are gaps or islands. Having an event
specifically say it is the `END` of a chain will allow for UIs or systems to
act accordingly based off the ending notation.

```json
{
  "context": {
    "version": "0.4.0-draft",
    "id": "76e96c6d-e418-42a5-b3f9-04d5c0699ac4",
    "chain_id": "7ff3f526-1a0e-4d35-8a4c-7d6295e97359",
    "source": "/event/source/123",
    "type": "dev.cdevents.build.queued.0.1.1",
    "timestamp": "2023-03-20T14:20:05.315384Z",
    "links": [
      {
        "link_type": "PATH",
        "from": {

>           "context_id": "5328c37f-bb7e-4bb7-84ea-9f5f85e4a7ce" # context id of a source.merged CDEvent

Reviewer: ```suggestion
          "context_id": "5328c37f-bb7e-4bb7-84ea-9f5f85e4a7ce" # context id of a change.merged CDEvent
```

        }
      }
    ]
  },
  "subject": {
    "id": "pkg:golang/mygit.com/myorg/myapp@234fd47e07d1004f0aed9c",
    "source": "/event/source/123",
    "type": "artifact",
    "content": {
      "change": {
        "id": "myChange123",
        "source": "my-git.example/an-org/a-repo"
      }
    }
  }
}
```

Above shows a simple path link that would allow for a connection between the
source changed and the build queued event. To illustrate links further, we can
allow for an artifact relation between what artifact was published to the
source change event which is shown below.

```json
{
  "context": {
    "version": "0.4.0-draft",
    "id": "271069a8-fc18-44f1-b38f-9d70a1695819",
    "chain_id": "7ff3f526-1a0e-4d35-8a4c-7d6295e97359",
    "source": "/event/source/123",
    "type": "dev.cdevents.artifact.published.0.1.1",
    "timestamp": "2023-03-20T14:35:06.123456Z",
    "links": [
      {
        "link_type": "RELATION",

>         "link_kind": "ARTIFACT",

Reviewer: This, to me, is not an artifact relation, but rather a "SOURCE_CHANGE" relation, meaning that this artifact being published is related to the source change previously informed about through the source merged event in the `target`.

Author: Hmm, depends on the point of view. I tend to look at it as this artifact is relating to the source change, so it's relation is the artifact. However, it could also be a source change relation. It's a matter of perspective in this regard. Why may want to dictate perspective or make it flexible like how we currently have it

Author: Ah gotcha, so saying it's an artifact relation is redundant. Hmm, let me see what works best

Author: But it is saying that the relation of artifact is to the source change, so I think it may be fine. It seems no different with source change since the source change event implies it as well.

When you are graphing something knowing what the artifact relates to could be useful, which was the intention here

Author: Ah gotcha, so basically due to the lack of information from the UUID, e.g. the context id, it doesn't provide too much useful information.

Yea, that makes sense. Okay, let me think a good relation type, `SOURCE_CHANGE` seems appropriate, but wonder if there's a more appropriate example

Reviewer: Some other examples could be to relate a pipelineRun.queued to a change.merged through a `TRIGGER` relation, or a build.started to an environment.created through an `ENVIRONMENT` relation, or an incident.detected to a service.deployed through a `SERVICE` relation. Just to mention a few of a great number of possible relations.

        "target": {

>           "context_id": "5328c37f-bb7e-4bb7-84ea-9f5f85e4a7ce"

Reviewer: ```suggestion
          "context_id": "5328c37f-bb7e-4bb7-84ea-9f5f85e4a7ce"  # context id of a change.merged CDEvent
```
